tests/drop_counters/test_drop_counters.py

```diff
@@ -234,36 +403,125 @@ def base_verification(discard_group, pkt, ptfadapter, duthost, combined_counter,
+        if acl_drops != PKT_NUMBER:
+            fail_msg = "ACL drop counter was not incremented on iface {}. DUT ACL counter == {}; Sent pkts == {}".format(
+                dut_iface, acl_drops, PKT_NUMBER
+            )
+            pytest.fail(fail_msg)
+        if not COMBINED_ACL_DROP_COUNTER:
+            ensure_no_l3_drops(duthost)
+            ensure_no_l2_drops(duthost)
     else:
         pytest.fail("Incorrect 'discard_group' specified. Supported values: 'L2' or 'L3'")
 
 
 def test_equal_smac_dmac_drop(ptfadapter, duthost, setup, tx_dut_ports, pkt_fields):
     """
-    @summary: Verify that packet with equal SMAC and DMAC is dropped and L2 drop cunter incremented
+    @summary: Verify that packet with equal SMAC and DMAC is dropped and L2 drop counter incremented
     """
     dut_iface, ptf_tx_port_id, dst_mac, src_mac = get_test_ports_info(ptfadapter, duthost, setup, tx_dut_ports)
 
-    log_pkt_params(dut_iface, dst_mac, dst_mac, pkt_fields["ip_dst"], pkt_fields["ip_src"])
+    log_pkt_params(dut_iface, dst_mac, dst_mac, pkt_fields["ipv4_dst"], pkt_fields["ipv4_src"])
 
     pkt = testutils.simple_tcp_packet(
         eth_dst=dst_mac, # DUT port
         eth_src=dst_mac, # PTF port
-        ip_src=pkt_fields["ip_src"], # PTF source
-        ip_dst=pkt_fields["ip_dst"], # DUT source
+        ip_src=pkt_fields["ipv4_src"], # PTF source
+        ip_dst=pkt_fields["ipv4_dst"], # VM source
         tcp_sport=pkt_fields["tcp_sport"],
         tcp_dport=pkt_fields["tcp_dport"])
 
-    base_verification("L2", pkt, ptfadapter, duthost, setup["combined_drop_counter"], ptf_tx_port_id, dut_iface)
+    base_verification("L2", pkt, ptfadapter, duthost, ptf_tx_port_id, dut_iface)
+
+    # Verify packets were not egresed the DUT
+    exp_pkt = expected_packet_mask(pkt)
+    testutils.verify_no_packet_any(ptfadapter, exp_pkt, ports=setup["neighbor_sniff_ports"])
+
+
+def test_multicast_smac_drop(ptfadapter, duthost, setup, tx_dut_ports, pkt_fields):
+    """
+    @summary: Verify that packet with multicast SMAC is dropped and L2 drop counter incremented
+    """
+    multicast_smac = "01:00:5e:00:01:02"
+    dut_iface, ptf_tx_port_id, dst_mac, src_mac = get_test_ports_info(ptfadapter, duthost, setup, tx_dut_ports)
+
+    log_pkt_params(dut_iface, dst_mac, multicast_smac, pkt_fields["ipv4_dst"], pkt_fields["ipv4_src"])
+
+    pkt = testutils.simple_tcp_packet(
+        eth_dst=dst_mac, # DUT port
+        eth_src=multicast_smac,
+        ip_src=pkt_fields["ipv4_src"], # PTF source
+        ip_dst=pkt_fields["ipv4_dst"], # VM source
+        tcp_sport=pkt_fields["tcp_sport"],
+        tcp_dport=pkt_fields["tcp_dport"])
+
+    base_verification("L2", pkt, ptfadapter, duthost, ptf_tx_port_id, dut_iface)
+
+    # Verify packets were not egresed the DUT
+    exp_pkt = expected_packet_mask(pkt)
+    testutils.verify_no_packet_any(ptfadapter, exp_pkt, ports=setup["neighbor_sniff_ports"])
```

Comment:
> This sequence is repeated in almost every test with slightly different parameters. Is there any way we can refactor this into one or more helper methods so that the setup and verification can be reused among all of the test cases?

I think it is worth to optimize sending packets and results verification, below code can be moved to some helper function:
```
    base_verification("L3", pkt, ptfadapter, duthost, ptf_tx_port_id, tx_dut_ports[dut_iface])

    # Verify packets were not egresed the DUT
    exp_pkt = expected_packet_mask(pkt)
    testutils.verify_no_packet_any(ptfadapter, exp_pkt, ports=setup["neighbor_sniff_ports"])
```

The rest is unique for every test case, and it will be more readable to keep it inside test case body.
What do you think?

Reply:
Fixed this.